### ISSUE TITLE
feat: Allow for new WebSocket URLs to be asynchronously injected on reconnect

### ISSUE THREAD

**Issue by spsDrop** [trpc/trpc#4188]
### Describe the feature you'd like to request

Reusing urls for WebSocket disconnects is not an option for me
- My WebSocket security setup requires that I use a pre-signed url that has been authed
- The url is only valid for one connection
- The url expires after 1 minute if not used to create a connection

I would like a way to asynchronously get a new url on re-connect. 

Right now I have to track all subscriptions and on disconnect remove them close the connection and re-create the proxy then re-re-add them after I create a new client with the updated url

### Describe the solution you'd like to see

I would like to provide an additional option parameter to `WebSocketClientOptions`
`getRetryUrl?: () =>  Promise<string>;`
And get the new url on reconnects

### Describe alternate solutions

Make the existing url parameter take an async function and lazily connect.

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR implementing this feature!

<sub>[TRP-21](https://linear.app/trpc/issue/TRP-21/feat-allow-for-new-websocket-urls-to-be-asynchronously-injected-on)</sub>

## Funding
* You can sponsor this specific effort via a [Polar.sh](https://polar.sh) pledge below
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/trpc/trpc/issues/4188">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/trpc/trpc/issues/4188/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/trpc/trpc/issues/4188/pledge.svg">
</picture>
</a>

**Comment by KATT** [comment]
I guess another solution would be to make `url` to `Promise<string>`?

```
export interface WebSocketClientOptions {
  url: string | (() => string | Promise<string>);
  // ...
}
```

**Comment by spsDrop** [comment]
> 

Yes that would be my ideal solution. But it would require more of a refactor to only create the WS connection on demand. The solution I implemented for now was the fastest to get us unblocked. Now that I have my local implementation working, I have more time to make the URL param async. It would change the behavior of the client pretty significantly, so I would like to make a sure that is the route we want to go with.

**Comment by KATT** [comment]
I'd love it to be done in the "proper" way but it needs to be done in a non-breaking way.

Idea:

1. Copy/paste `createClient()` to `createNewClient()` (won't impact bundle sizes b/c of tree shaking
2. Fix `createNewClient` 
3. Make `httpLink` work with either interface
4. Merge this to `main`

--- 

5. Create new PR to `next` where we delete the old `createClient` and rename `createNewClient` to `createClient`
6. Make a note in https://github.com/trpc/trpc/blob/next/.wip/changes.md

**Comment by spsDrop** [comment]
That makes sense but I was thinking I could fix this directly within `wsLink.ts`. So the change would be create an `asyncWsLink.ts` that lazily creates the ws on first dispatch. That way we don't have to make changes to httpLink. In fact this would make the wsLink function more like the httpLink because it `headers()` is can be an async function.

**Comment by spsDrop** [comment]
Initial work PR up here:
https://github.com/trpc/trpc/pull/4205

**Comment by spsDrop** [comment]
This is in v11. Thank you!
